### PR TITLE
Add LlamaIndex RAG

### PR DESCRIPTION
**Pull request by ivy-lv11**
# LlamaIndex RAG Pipeline Example
* Add llamaindex RAG example with bigdl
* Add llamaindex unit test
* Add support for llamaindex in src

### CODE REVIEW

File: python/llm/src/bigdl/llm/llamaindex/llms/bigdlllm.py [hunk 1]
```diff
@@ -0,0 +1,325 @@
+import logging
```
Comment (shane-huang) [inline python/llm/src/bigdl/llm/llamaindex/llms/bigdlllm.py:1]: add license

Comment (ivy-lv11) [reply python/llm/src/bigdl/llm/llamaindex/llms/bigdlllm.py:47]: License is added now.

File: python/llm/src/bigdl/llm/llamaindex/llms/bigdlllm.py [hunk 2]
```diff
@@ -0,0 +1,371 @@
...
+# limitations under the License.
+#
+
+# This would makes sure Python is aware there is more than one sub-package within bigdl,
+# physically located elsewhere.
+# Otherwise there would be module not found error in non-pip's setting as Python would
+# only search the first bigdl package and end up finding only one sub-package.
+
+# This file is adapted from
+# https://github.com/hwchase17/langchain/blob/master/langchain/llms/llamacpp.py
```
Comment (shane-huang) [inline python/llm/src/bigdl/llm/llamaindex/llms/bigdlllm.py:23]: langchain? should be some llamaindex file?

File: python/llm/src/bigdl/llm/llamaindex/llms/bigdlllm.py [hunk 3]
```diff
@@ -0,0 +1,371 @@
...
+        generate_kwargs: Optional[dict] = None,
+        is_chat_model: Optional[bool] = False,
+        callback_manager: Optional[CallbackManager] = None,
+        system_prompt: str = "",
+        messages_to_prompt: Optional[Callable[[Sequence[ChatMessage]], str]]=None,
+        completion_to_prompt: Optional[Callable[[str], str]]=None,
+        pydantic_program_mode: PydanticProgramMode=PydanticProgramMode.DEFAULT,
+        output_parser: Optional[BaseOutputParser] = None,
+    ) -> None:
+        """Initialize params."""
```
Comment (shane-huang) [inline python/llm/src/bigdl/llm/llamaindex/llms/bigdlllm.py:200]: add api docs for automatically api doc generation

File: python/llm/src/bigdl/llm/llamaindex/llms/bigdlllm.py [hunk 4]
```diff
@@ -0,0 +1,371 @@
...
+        model_kwargs = model_kwargs or {}
+        from bigdl.llm.transformers import AutoModelForCausalLM
+        self._model = model or AutoModelForCausalLM.from_pretrained(
+            model_name, load_in_4bit=True, **model_kwargs
+        )
+        if device_map == 'xpu':
```
Comment (shane-huang) [inline python/llm/src/bigdl/llm/llamaindex/llms/bigdlllm.py:206]: did you verify this check on GPU?

File: python/llm/test/llamaindex/test_llamaindex.py [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+
+            # add final assistant prompt
+            prompt = prompt + "<|assistant|>\n"
+            return prompt
+        
+        llm = BigdlLLM(
+            model_name=self.llama_model_path,
+            tokenizer_name=self.llama_model_path,
+            context_window=3900,
+            max_new_tokens=256,
```
Comment (shane-huang) [inline python/llm/test/llamaindex/test_llamaindex.py:75]: use a smaller max_new_tokens for tests to finish faster. do we need a large context_windows like 3900?

File: python/llm/test/llamaindex/test_llamaindex.py [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+        self.llama_model_path = os.environ.get('LLAMA_ORIGIN_PATH')
+        self.bloom_model_path = os.environ.get('BLOOM_ORIGIN_PATH')
+        thread_num = os.environ.get('THREAD_NUM')
+        if thread_num is not None:
+            self.n_threads = int(thread_num)
+        else:
+            self.n_threads = 2     
+    
+    def test_bigdl_llm(self):    
+        def completion_to_prompt(completion):
```
Comment (shane-huang) [inline python/llm/test/llamaindex/test_llamaindex.py:51]: take `completion_to_prompt` and `messages_to_prompt` out of  `test_bigdl_llm` and make them methods of `Test_Langchain_Transformers_API`.

File: python/llm/src/bigdl/llm/llamaindex/llms/bigdlllm.py (hunk 2 above)
Comment (ivy-lv11) [reply python/llm/src/bigdl/llm/llamaindex/llms/bigdlllm.py:23]: Yes, it should be llamaindex related files. I will fix it.

File: python/llm/test/llamaindex/test_llamaindex.py [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+            tokenizer_name=self.llama_model_path,
+            context_window=512,
+            max_new_tokens=32,
+            model_kwargs={},
+            generate_kwargs={"temperature": 0.7, "do_sample": False},
+            messages_to_prompt=self.messages_to_prompt,
+            completion_to_prompt=self.completion_to_prompt,
+            device_map="cpu",
+        )
+        res = llm.complete("What is AI?")
```
Comment (shane-huang) [inline python/llm/test/llamaindex/test_llamaindex.py:83]: add assert to test if `res` is empty. refer to other unit tests and check how they did it.

File: python/llm/src/bigdl/llm/llamaindex/llms/bigdlllm.py [hunk 8]
```diff
@@ -0,0 +1,452 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+# The file is modified from
+# BigDL/python/llm/src/bigdl/llm/langchain/llms/bigdlllm.py
```
Comment (shane-huang) [inline python/llm/src/bigdl/llm/llamaindex/llms/bigdlllm.py:18]: you don't need to refer to our own lib. Just files from llama index. And add the full hyperlink for `base.py`

File: python/llm/example/CPU/LlamaIndex/README.md [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+- `-m MODEL_PATH`: **required**, path to the llama model
+- `-e EMBEDDING_MODEL_PATH`: path to the embedding model
+- `-u USERNAME`: username in the postgres database
+- `-p PASSWORD`: password in the postgres database
+- `-q QUESTION`: question you want to ask
+- `-d DATA`: path to data used during retrieval
+
+Here is the sample output when applying Llama-2-7b-chat-hf as the generatio model. 
+```
+AI stands for Artificial Intelligence. It refers to the development of computer systems that can perform tasks that typically require human intelligence, such as understanding natural language, recognizing images, making decisions, and solving problems. AI systems use algorithms and machine learning techniques to analyze data and make predictions or decisions based on that data.
```
Comment (shane-huang) [inline python/llm/example/CPU/LlamaIndex/README.md:59]: Use the output of the original question:  "How does Llama 2 perform compared to other open-source models?" as example output. We need to demonstrate that it answers the question out of query from the input llama2 paper pdf

**Comment by shane-huang** [comment]
merge for now and revise readme in following PR.

**Review by shane-huang — APPROVED** [review]
